generic_server.cc

```diff
@@ -9,19 +9,112 @@
 #include "generic_server.hh"
 
 
+#include <exception>
 #include <fmt/ranges.h>
 #include <seastar/core/when_all.hh>
 #include <seastar/coroutine/parallel_for_each.hh>
 #include <seastar/core/reactor.hh>
 #include <seastar/core/smp.hh>
+#include <utility>
+
+#include "db/config.hh"
 
 namespace generic_server {
 
-connection::connection(server& server, connected_socket&& fd)
-    : _server{server}
+class counted_data_source_impl : public data_source_impl {
+    data_source _ds;
+    connection::cpu_concurrency_t& _cpu_concurrency;
+
+    template <typename F>
+    future<temporary_buffer<char>> invoke_with_counting(F&& fun) {
+        if (_cpu_concurrency.stopped) {
+            return fun();
+        }
+        return futurize_invoke([this] () {
+            _cpu_concurrency.units.return_all();
+        }).then([fun = std::move(fun)] () {
+            return fun();
+        }).finally([this] () {
+            _cpu_concurrency.units.adopt(consume_units(_cpu_concurrency.semaphore, 1));
```

Comment:
I don't understand this.
We return the units before calling fun() and then consume them again after fun() ran, which will prevent new connections.  Was that the intention?

Reply:
yes, fun here represents network IO, so before we "mark" it as CPU phase and limit, then in IO we don't limit, and after fun we go back to CPU limiting phase

Reply:
> But if processing which is async blocks on other things, we would still throttle.

Yes that's the simplification here. Alternative would be tagging the code everywhere and unlikely to improve the mechanism much for real scenario.

> therefore shedding is more likely causing the test to be extra flaky

You mean some concrete examples or in theory? This could only be the case if tests themselves are making lots of connections, or there is very high parallelism in testing with cluster reuse. To me it looks like both are unlikely (I think test.py runs just couple tests in parallel) but yes, in theory can happen.